daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -347,9 +347,66 @@ object Value extends CidContainer1WithDefaultCidResolver[Value] {
     * to be able to use AbsoluteContractId elsewhere, so that we can
     * automatically upcast to ContractId by subtyping.
     */
-  sealed trait ContractId
+  sealed abstract class ContractId
+
+  sealed abstract class AbsoluteContractId extends ContractId with Product with Serializable {
+    def coid: String
+  }
+
+  object AbsoluteContractId {
+    final case class V0(coid: Ref.ContractIdString) extends AbsoluteContractId {
+      override def toString: String = s"AbsoluteContractId($coid)"
+    }
+
+    object V0 {
+      def fromString(s: String): Either[String, V0] =
+        Ref.ContractIdString.fromString(s).map(V0(_))
+
+      def assertFromString(s: String): V0 = assertRight(fromString(s))
+    }
+
+    final case class V1(discriminator: crypto.Hash, suffix: Bytes = Bytes.Empty)
+        extends AbsoluteContractId {
+      lazy val toBytes: Bytes = V1.prefix ++ discriminator.bytes ++ suffix
+      lazy val coid: Ref.HexString = toBytes.toHexString
```

Comment:
Often a contractId is bound to a name like `coid`. Referencing the hex string the looks like `coid.coid`. How about we call this `toHexString`? Also makes it very explicit what this is.

Reply:
:+1:

Reply:
We need the same name for V0 and V1. V0 coid is not a `HexString`.

Reply:
I definitively do not want to use `toString`. Any suggestion ?

Reply:
okay, let's keep it for now.